pkg/continuoustest/write_read_series.go

```diff
@@ -96,9 +97,93 @@ func (t *WriteReadSeriesTest) Run(ctx context.Context, now time.Time) {
 		t.lastWrittenTimestamp = timestamp
+		t.queryMaxTime = timestamp
+		if t.queryMinTime.IsZero() {
+			t.queryMinTime = timestamp
+		}
+	}
+
+	for _, timeRange := range t.getRangeQueryTimeRanges(now) {
+		t.runRangeQueryAndVerifyResult(ctx, timeRange[0], timeRange[1])
+	}
+}
+
+// getRangeQueryTimeRanges returns the start/end time ranges to use to run test range queries.
+func (t *WriteReadSeriesTest) getRangeQueryTimeRanges(now time.Time) (ranges [][2]time.Time) {
+	// The min and max allowed query timestamps are zero if there's no successfully written data yet.
+	if t.queryMinTime.IsZero() || t.queryMaxTime.IsZero() {
+		level.Info(t.logger).Log("msg", "Skipped range queries because there's no valid time range to query")
+		return nil
+	}
+
+	// Honor the configured max age.
+	adjustedQueryMinTime := maxTime(t.queryMinTime, now.Add(-t.cfg.MaxQueryAge))
+	if t.queryMaxTime.Before(adjustedQueryMinTime) {
+		level.Info(t.logger).Log("msg", "Skipped range queries because there's no valid time range to query after honoring configured max query age", "min_valid_time", t.queryMinTime, "max_valid_time", t.queryMaxTime, "max_query_age", t.cfg.MaxQueryAge)
+		return
+	}
+
+	// Last 1h.
+	if t.queryMaxTime.After(now.Add(-1 * time.Hour)) {
+		ranges = append(ranges, [2]time.Time{
+			maxTime(adjustedQueryMinTime, now.Add(-1*time.Hour)),
+			minTime(t.queryMaxTime, now),
+		})
+	}
+
+	// Last 24h (only if the actual time range is not already covered by "Last 1h").
+	if t.queryMaxTime.After(now.Add(-24*time.Hour)) && adjustedQueryMinTime.Before(now.Add(-1*time.Hour)) {
+		ranges = append(ranges, [2]time.Time{
+			maxTime(adjustedQueryMinTime, now.Add(-24*time.Hour)),
+			minTime(t.queryMaxTime, now),
+		})
 	}
 
-	// TODO Here we should query the written data and assert on correctness.
+	// From last 23h to last 24h.
+	if adjustedQueryMinTime.Before(now.Add(-23*time.Hour)) && t.queryMaxTime.After(now.Add(-23*time.Hour)) {
+		ranges = append(ranges, [2]time.Time{
+			maxTime(adjustedQueryMinTime, now.Add(-24*time.Hour)),
+			minTime(t.queryMaxTime, now.Add(-23*time.Hour)),
+		})
+	}
+
+	// A random time range.
+	randMinTime := randTime(adjustedQueryMinTime, t.queryMaxTime)
+	ranges = append(ranges, [2]time.Time{randMinTime, randTime(randMinTime, t.queryMaxTime)})
+
+	return ranges
+}
+
+func (t *WriteReadSeriesTest) runRangeQueryAndVerifyResult(ctx context.Context, start, end time.Time) {
+	// We align start, end and step to write interval in order to avoid any false positives
+	// when checking results correctness. The min/max query time is always aligned.
+	start = maxTime(t.queryMinTime, alignTimestampToInterval(start, writeInterval))
+	end = minTime(t.queryMaxTime, alignTimestampToInterval(end, writeInterval))
+	if end.Before(start) {
+		return
+	}
+
+	step := getQueryStep(start, end, writeInterval)
+	query := fmt.Sprintf("sum(%s)", metricName)
+
+	logger := log.With(t.logger, "query", query, "start", start.Unix(), "end", end.Unix(), "step", step)
```

Comment:
Nit: Here we log timestamp in unix seconds, but in util.go we log unix milliseconds.